docs/core/architecture.md

```diff
@@ -38,33 +40,79 @@ Namespaces are designated with "\".
-on the data schema.  DAO objects tend to be instantiated in BAO classes.
+The CiviCRM **data access objects** (DAOs) are PHP classes that
+([e.g. `CRM/Pledge/DAO`](https://github.com/civicrm/civicrm-core/tree/master/CRM/Pledge/DAO))
+expose the contents
+of the database.  The release script generates each DAO automatically based
+on the matching XML file in the [data schema](#database-structure).  DAO objects tend to be instantiated in BAO classes.
+
+The DAO classes all extend the core
+[DAO base class](https://github.com/civicrm/civicrm-core/blob/master/CRM/Core/DAO.php)
+which itself is an extension of the external
+[DataObject class](https://github.com/civicrm/civicrm-packages/blob/master/DB/DataObject.php).
+These base classes provide standard CRUD (create, retrieve, update and delete)
+methods, etc. <!--fixme why the etc? what else?? -->
+
+The generated DAO object has:
+
+* A property for each field using the actual field name, not the unique name
+* A `links()` method which retrieves the links to other tables (off the foreign keys)
+* An `import()` method and an `export()` method for ?
+* A `fields()` method which returns an array of fields for that object keyed by the field's unique name. Looking at the field 'pledge.amount' we see
+* A couple of functions to define the labels for enumerated fields
+
+```php
+  'pledge_amount' => array(
+    'name' => 'amount',
+    'type' => CRM_Utils_Type::T_MONEY,
+    'title' => ts('Total Pledged') ,
+    'required' => true,
+    'import' => true,
+    'where' => 'civicrm_pledge.amount',
+    'headerPattern' => '',
+    'dataPattern' => '',
+    'export' => true,
+    'bao' => 'CRM_Pledge_BAO_Pledge',
+    'table_name' => 'civicrm_pledge',
+    'entity' => 'Pledge',
+  ),
+```
+
+The key is the unique name but the name field is the field's name and the 'where' field shows the MySQL description of it. We also see the data type and whether it is available for search or export.
 
-The DAO has a property for each field (using the actual field name, not the
-unique name).  They also have standard CRUD (create retrieve update delete) type
-functions, etc. <!--fixme why the etc? what else?? -->
+Generally fields should be exportable unless there is a security reason or they are weird and confusing as the search builder is also driven by this setting.
+
+Fields whose option values can be calculated will also have a `pseudoconstant` section.
 
 ### BAO
-BAO stands for business access object.  BAOs map to DAOs and extend them with
+BAO stands for business access object
+([example](https://github.com/civicrm/civicrm-core/blob/master/CRM/Event/BAO/Event.php)).
+BAOs map to DAOs and extend them with
 the business logic of CiviCRM.  A lot of the meat of CiviCRM is found in the
 BAOs, for example they have the code that creates follow up activities when an
 activity is created, or create activities and populating custom fields when a
 pledge is created.
 
+!!! note
```

Comment:
This is incorrect - they are really really inconsistent - I would probably argue we should move towards removing all the add functions & only having create